--- a/rt2860v2/ap/apcli_ctrl.c
+++ b/rt2860v2/ap/apcli_ctrl.c
@@ -492,6 +492,7 @@ static VOID ApCliCtrlJoinReqTimeoutActio
 	{
 		COPY_MAC_ADDR(JoinReq.Bssid, pApCliEntry->CfgApCliBssid);
 	}
+	pApCliEntry->ProbeTimeoutCount++;
 
 #ifdef WSC_AP_SUPPORT
     if ((pAd->ApCfg.ApCliTab[ifIndex].WscControl.WscConfMode != WSC_DISABLE) &&
--- a/rt2860v2/ap/apcli_sync.c
+++ b/rt2860v2/ap/apcli_sync.c
@@ -385,7 +385,7 @@ static VOID ApCliPeerProbeRspAtJoinActio
 				
 				pApCliEntry->ApCliMlmeAux.VarIELen = 0;
 			}
-
+			pApCliEntry->ProbeTimeoutCount = 0;
 			DBGPRINT(RT_DEBUG_TRACE, ("SYNC - receive desired PROBE_RSP at JoinWaitProbeRsp... Channel = %d\n", Channel));
 
 			/* if the Bssid doesn't be indicated then you need to decide which AP to connect by most strong Rssi signal strength. */
@@ -660,6 +660,8 @@ static VOID ApCliEnqueueProbeRequest(
 	BOOLEAN bHasWscIe = FALSE;
 #endif
 
+	BOOLEAN       NeedBroadcast = FALSE;
+
 	DBGPRINT(RT_DEBUG_TRACE, ("force out a ProbeRequest ...\n"));
 
 	if (ifIndex >= MAX_APCLI_NUM)
@@ -674,12 +676,31 @@ static VOID ApCliEnqueueProbeRequest(
 		return;
 	} else
 	{
-		if(MAC_ADDR_EQUAL(pAd->ApCfg.ApCliTab[ifIndex].CfgApCliBssid, ZERO_MAC_ADDR))
-			ApCliMgtMacHeaderInit(pAd, &Hdr80211, SUBTYPE_PROBE_REQ, 0,
-				BROADCAST_ADDR, BROADCAST_ADDR, ifIndex);
+		if(pAd->ApCfg.ApCliTab[ifIndex].ProbeTimeoutCount < DEFAULT_PROBE_BROADCAST)
+		{
+			NeedBroadcast = FALSE;
+		}
 		else
+		{
+			NeedBroadcast = TRUE;
+			pAd->ApCfg.ApCliTab[ifIndex].ProbeTimeoutCount = 0;
+		}
+		
+		if(NeedBroadcast == TRUE)
+		{
+			//DBGPRINT(RT_DEBUG_ERROR, ("APCLI: Will try  connect iPhone 's Hotspot!!\n", __FUNCTION__));
 			ApCliMgtMacHeaderInit(pAd, &Hdr80211, SUBTYPE_PROBE_REQ, 0,
-				pAd->ApCfg.ApCliTab[ifIndex].CfgApCliBssid, pAd->ApCfg.ApCliTab[ifIndex].CfgApCliBssid, ifIndex);
+					BROADCAST_ADDR, BROADCAST_ADDR, ifIndex);
+		}
+		else
+		{
+			if(MAC_ADDR_EQUAL(pAd->ApCfg.ApCliTab[ifIndex].CfgApCliBssid, ZERO_MAC_ADDR))
+				ApCliMgtMacHeaderInit(pAd, &Hdr80211, SUBTYPE_PROBE_REQ, 0,
+					BROADCAST_ADDR, BROADCAST_ADDR, ifIndex);
+			else
+				ApCliMgtMacHeaderInit(pAd, &Hdr80211, SUBTYPE_PROBE_REQ, 0,
+					pAd->ApCfg.ApCliTab[ifIndex].CfgApCliBssid, pAd->ApCfg.ApCliTab[ifIndex].CfgApCliBssid, ifIndex);
+		}
 
 		ssidLen = SsidLen;
 		NdisZeroMemory(ssid, MAX_LEN_OF_SSID);
--- a/rt2860v2/include/rtmp.h
+++ b/rt2860v2/include/rtmp.h
@@ -2926,6 +2926,7 @@ typedef struct _APCLI_COUNTER {
 	ULONG MulticastReceivedFrameCount;
 } APCLI_COUNTER, *PAPCLI_COUNTER;
 
+#define DEFAULT_PROBE_BROADCAST 5
 typedef struct _APCLI_STRUCT {
 	PNET_DEV dev;
 #ifdef RTL865X_SOC
@@ -3045,7 +3046,8 @@ typedef struct _APCLI_STRUCT {
 
 	PSPOLL_FRAME PsPollFrame;
 	HEADER_802_11 NullFrame;
-
+	
+	UCHAR ProbeTimeoutCount;
 	UAPSD_INFO	UapsdInfo;
 #ifdef MAC_REPEATER_SUPPORT
 	REPEATER_CLIENT_ENTRY RepeaterCli[MAX_EXT_MAC_ADDR_SIZE];
