--- a/mt7610/os/linux/cfg80211drv.c
+++ b/mt7610/os/linux/cfg80211drv.c
@@ -844,7 +844,10 @@ BOOLEAN CFG80211_SupBandReInit(
 {
 	PRTMP_ADAPTER pAd = (PRTMP_ADAPTER)pAdCB;
 	CFG80211_BAND BandInfo;
-
+	CFG80211_CB *pCfg80211_CB = CFG80211CB;
+	struct wiphy *pWiphy = pCfg80211_CB->pCfg80211_Wdev->wiphy;
+ 
+	memcpy(pWiphy->perm_addr, pAd->PermanentAddress, ETH_ALEN);
 
 	CFG80211DBG(RT_DEBUG_ERROR, ("80211> re-init bands...\n"));
 
--- a/mt7610/os/linux/rt_linux.c
+++ b/mt7610/os/linux/rt_linux.c
@@ -1462,6 +1462,7 @@ int RtmpOSNetDevAttach(
 /*		GET_PAD_FROM_NET_DEV(pAd, pNetDev); */
 
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,31)
+		pNetDevOps->ndo_set_mac_address = eth_mac_addr;
 		pNetDevOps->ndo_open = pDevOpHook->open;
 		pNetDevOps->ndo_stop = pDevOpHook->stop;
 		pNetDevOps->ndo_start_xmit =
