--- a/mt76x2/mcu/mcu_and.c
+++ b/mt76x2/mcu/mcu_and.c
@@ -123,20 +123,20 @@ loadfw_protect:
 #ifdef DBG
 	DBGPRINT(RT_DEBUG_OFF, ("\n"));
 
-	DBGPRINT(RT_DEBUG_OFF, ("platform = "));
+	DBGPRINT(RT_DEBUG_OFF, ("platform = ["));
 	for (loop = 0; loop < 4; loop++)
 		DBGPRINT(RT_DEBUG_OFF, ("%c", *(cap->rom_patch + 16 + loop)));
-	DBGPRINT(RT_DEBUG_OFF, ("\n"));
+	DBGPRINT(RT_DEBUG_OFF, ("]\n"));
 
-	DBGPRINT(RT_DEBUG_OFF, ("hw/sw version = "));
+	DBGPRINT(RT_DEBUG_OFF, ("hw/sw version = ["));
 	for (loop = 0; loop < 4; loop++)
 		DBGPRINT(RT_DEBUG_OFF, ("%x", *(cap->rom_patch + 20 + loop)));
-	DBGPRINT(RT_DEBUG_OFF, ("\n"));
+	DBGPRINT(RT_DEBUG_OFF, ("]\n"));
 
-	DBGPRINT(RT_DEBUG_OFF, ("patch version = "));
+	DBGPRINT(RT_DEBUG_OFF, ("patch version = ["));
 	for (loop = 0; loop < 4; loop++)
-		DBGPRINT(RT_DEBUG_OFF, ("%x", *(cap->rom_patch + 24 + loop)));
-	DBGPRINT(RT_DEBUG_OFF, ("\n"));
+		DBGPRINT(RT_DEBUG_OFF, ("%02x", *(cap->rom_patch + 24 + loop)));
+	DBGPRINT(RT_DEBUG_OFF, ("]\n"));
 #endif
 
 	RTMP_IO_WRITE32(ad, PCIE_REMAP_BASE4, cap->rom_patch_offset - 10000);
