--- a/mt76x2/chips/mt76x2.c
+++ b/mt76x2/chips/mt76x2.c
@@ -1,5 +1,6 @@
 #include "rt_config.h"
 
+#ifndef LOAD_FW_FORM_FILES
 #include "mcu/mt7662_rom_patch.h"
 #include "mcu/mt7662_firmware.h"
 
@@ -7,6 +8,7 @@
 #include "mcu/mt7662t_rom_patch.h"
 #include "mcu/mt7662t_firmware.h"
 #endif /* MT7662TU */
+#endif
 
 #include "eeprom/mt76x2_e2p.h"
 
@@ -5785,6 +5787,15 @@ static const RTMP_CHIP_CAP MT76x2_ChipCa
 #ifdef RTMP_PCI_SUPPORT
 	.cmd_padding_len = 0,
 #endif
+#ifdef LOAD_FW_FORM_FILES
+	.fw_header_image = NULL,
+	.fw_len = 0,
+	.rom_patch_header_image = NULL,
+	.rom_patch_len = 0,
+	.fw_bin_file_name = "mt7662_firmware.bin",
+	.rom_patch_bin_file_name = "mt7662_patch_hdr.bin",
+	.load_code_method = BIN_FILE_METHOD,
+#else
 	.fw_header_image = MT7662_FirmwareImage,
 	.fw_bin_file_name = "mtk/WIFI_RAM_CODE_ALL.bin",
 	.fw_len = sizeof(MT7662_FirmwareImage),
@@ -5792,6 +5803,7 @@ static const RTMP_CHIP_CAP MT76x2_ChipCa
 	.rom_patch_len = sizeof(mt7662_rom_patch),
 	.rom_patch_bin_file_name = "mtk/mt7662_patch_e1_hdr.bin",
 	.load_code_method = HEADER_METHOD,
+#endif
 	.MACRegisterVer = "",
 	.BBPRegisterVer = "MT7662E1_BBP_CR_20130116.xls",
 	.RFRegisterVer = "",
--- a/mt76x2_ap/Makefile
+++ b/mt76x2_ap/Makefile
@@ -376,6 +376,7 @@ EXTRA_CFLAGS += -DRTMP_TEMPERATURE_TX_AL
 #EXTRA_CFLAGS += -DSPECIFIC_TX_POWER_SUPPORT
 #EXTRA_CFLAGS += -DTHERMAL_PROTECT_SUPPORT
 #EXTRA_CFLAGS += -DDMA_BUSY_RESET
+EXTRA_CFLAGS += -DLOAD_FW_FORM_FILES
 ifeq ($(CHIP_MT7620),y)
 EXTRA_CFLAGS += -DMULTI_INF_SUPPORT
 endif
