--- a/mt76x3/common/cmm_info.c
+++ b/mt76x3/common/cmm_info.c
@@ -2568,6 +2568,8 @@ VOID RTMPIoctlGetSiteSurvey(
     INT         max_len = LINE_LEN;		
 	BSS_ENTRY *pBss;
 	UINT32 TotalLen, BufLen = IW_SCAN_MAX_DATA;
+	BSS_TABLE *pScanTab;
+	pScanTab = &pAdapter->ScanTab;
 #ifdef AIRPLAY_SUPPORT
 	UCHAR TargetSsid[MAX_LEN_OF_SSID+1];
 	UCHAR TargetSsidLen = 0;
@@ -2628,6 +2630,8 @@ VOID RTMPIoctlGetSiteSurvey(
 
 	WaitCnt = 0;
 
+	BssTableSortByRssi(pScanTab,FALSE);
+
 	while ((ScanRunning(pAdapter) == TRUE) && (WaitCnt++ < 200))
 		OS_WAIT(500);	
 
