--- a/mt76x3/common/cmm_data_pci.c
+++ b/mt76x3/common/cmm_data_pci.c
@@ -31,7 +31,7 @@
 #include	"rt_config.h"
 
 
-#ifdef DBG
+#if 1
 VOID dump_txd(RTMP_ADAPTER *pAd, TXD_STRUC *pTxD)
 {
 	DBGPRINT(RT_DEBUG_OFF, ("TxD:\n"));
--- a/mt76x3/include/rt_config.h
+++ b/mt76x3/include/rt_config.h
@@ -272,5 +272,9 @@
 #include "hotspot.h"
 #endif
 
+#ifdef BG_FT_SUPPORT
+#include "br_ftph.h"
+#endif /* BG_FT_SUPPORT */
+
 #endif	/* __RT_CONFIG_H__ */
 
--- a/mt76x3/os/linux/br_ftph.c
+++ b/mt76x3/os/linux/br_ftph.c
@@ -86,6 +86,7 @@ Note:
 */
 VOID BG_FTPH_Init(VOID)
 {
+	printk("MT7603E Fast Mode init.\n");
 	RALINK_FP_Handle = BG_FTPH_PacketFromApHandle;
 }
 
@@ -107,6 +108,7 @@ Note:
 */
 VOID BG_FTPH_Remove(VOID)
 {
+	printk("MT7603E Fast Mode exit.\n");
 	RALINK_FP_Handle = NULL;
 } /* End of BG_FTPH_Init */
 
@@ -144,9 +146,21 @@ UINT32 BG_FTPH_PacketFromApHandle(
 
 	if (pNetDev != NULL)
 	{
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,31)
+		if (br_port_get_rcu(pNetDev) != NULL)
+#else
 		if (pNetDev->br_port != NULL)
+#endif
 		{
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,12)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,10,0)
+			/* br_fdb_get is not exported symbol, need exported in net/bridge/br.c */
+			pDstFdbEntry = __br_fdb_get(br_port_get_rcu(pNetDev)->br, pRxPkt->data, 0);
+			pSrcFdbEntry = __br_fdb_get(br_port_get_rcu(pNetDev)->br, pRxPkt->data + 6, 0);
+#elif  LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,31)
+			/* br_fdb_get is not exported symbol, need exported in net/bridge/br.c */
+			pDstFdbEntry = __br_fdb_get(br_port_get_rcu(pNetDev)->br, pRxPkt->data);
+			pSrcFdbEntry = __br_fdb_get(br_port_get_rcu(pNetDev)->br, pRxPkt->data + 6);
+#elif LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,12)
 			pDstFdbEntry = br_fdb_get_hook(pNetDev->br_port->br, pRxPkt->data);
 			pSrcFdbEntry = br_fdb_get_hook(pNetDev->br_port->br, pRxPkt->data + 6);
 #else
@@ -162,7 +176,11 @@ UINT32 BG_FTPH_PacketFromApHandle(
 				(pDstFdbEntry->dst == NULL) ||
 				(pDstFdbEntry->dst->dev == NULL) ||
 				(pDstFdbEntry->dst->dev == pNetDev) ||
+#if  LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,31)
+				(br_port_get_rcu(pNetDev)->state != BR_STATE_FORWARDING) ||
+#else
 				(pNetDev->br_port->state != BR_STATE_FORWARDING) ||
+#endif
 				((pSrcFdbEntry->dst != NULL) &&
 					(pSrcFdbEntry->dst->dev != NULL) &&
 					(pSrcFdbEntry->dst->dev != pNetDev)))
@@ -176,7 +194,11 @@ UINT32 BG_FTPH_PacketFromApHandle(
 				(pDstFdbEntry->dst->dev != NULL))
 			{
 				pRxPkt->dev = pDstFdbEntry->dst->dev;
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,31)
 				pDstFdbEntry->dst->dev->hard_start_xmit(pRxPkt, pDstFdbEntry->dst->dev);
+#else
+				pDstFdbEntry->dst->dev->netdev_ops->ndo_start_xmit(pRxPkt, pDstFdbEntry->dst->dev);
+#endif
 				return 0;
 			} /* End of if */
 		} /* End of if */
--- a/mt76x3_ap/Makefile
+++ b/mt76x3_ap/Makefile
@@ -88,6 +88,9 @@ $(RT_DRV_NAME)-objs += $(RT_DRV_PATH)/os
 $(RT_DRV_NAME)-objs += $(RT_DRV_PATH)/os/linux/rt_pci_rbus.o
 #$(RT_DRV_NAME)-objs += $(RT_DRV_PATH)/os/linux/rt_txrx_hook.o
 $(RT_DRV_NAME)-objs += $(RT_DRV_PATH)/os/linux/pci_main_dev.o
+ifeq ($(CONFIG_RT_BRIDGE_FASTPATH),y)
+$(RT_DRV_NAME)-objs += $(RT_DRV_PATH)/os/linux/br_ftph.o
+endif
 
 ### MT7603
 $(RT_DRV_NAME)-objs += $(RT_DRV_PATH)/chips/rtmp_chip.o
@@ -458,6 +461,10 @@ ifeq ($(CONFIG_RT_MC_SUPPORT),y)
 EXTRA_CFLAGS += -DMULTIPLE_CARD_SUPPORT
 endif
 
+ifeq ($(CONFIG_RT_BRIDGE_FASTPATH),y)
+EXTRA_CFLAGS += -DBG_FT_SUPPORT
+endif
+
 #EXTRA_CFLAGS += -DCONFIG_RT_FIRST_CARD=$(CONFIG_RT_FIRST_CARD)
 EXTRA_CFLAGS += -DCONFIG_RT_FIRST_CARD=7603
 
