diff -Nru a/mt76x3/rate_ctrl/alg_grp.c b/mt76x3/rate_ctrl/rate_ctrl/alg_grp.c
--- a/mt76x3/rate_ctrl/alg_grp.c	2016-05-26 17:22:22.000000000 +0800
+++ b/mt76x3/rate_ctrl/alg_grp.c	2016-08-18 11:59:04.478048600 +0800
@@ -952,8 +952,29 @@
 
 		if (CurrRateIdx != DownRateIdx)
 		{
-			pEntry->CurrTxRateIndex = DownRateIdx;
-			pEntry->LastSecTxRateChangeAction = RATE_DOWN;
+			if(pAd->bPandoraBoxSmartEn) {
+						UCHAR *pTable = pEntry->pTable;
+						CHAR Rssi;
+						RTMP_RA_LEGACY_TB *pCurrTxRate;
+			
+						Rssi = RTMPAvgRssi(pAd, &pEntry->RssiSample);
+						pCurrTxRate = PTX_RA_LEGACY_ENTRY(pTable, CurrRateIdx);
+			
+						if(Rssi > -60 && pCurrTxRate->CurrMCS == 3)
+						{
+							pEntry->LastSecTxRateChangeAction = RATE_NO_CHANGE;
+							printk("%s;%d;fix lowest mcs=3;pCurrTxRate->CurrMCS=%d=\n",__FUNCTION__,__LINE__,pCurrTxRate->CurrMCS);
+						}
+						else
+						{
+						pEntry->CurrTxRateIndex = DownRateIdx;
+						pEntry->LastSecTxRateChangeAction = RATE_DOWN;
+						}
+			} else {
+						pEntry->CurrTxRateIndex = DownRateIdx;
+						pEntry->LastSecTxRateChangeAction = RATE_DOWN;
+			}
+
 		}
 	}
 	else
diff -Nru a/mt76x3/rate_ctrl/alg_legacy.c b/mt76x3/rate_ctrl/alg_legacy.c
--- a/mt76x3/rate_ctrl/alg_legacy.c	2016-05-26 17:22:22.000000000 +0800
+++ b/mt76x3/rate_ctrl/alg_legacy.c	2016-08-18 12:15:32.373048600 +0800
@@ -806,8 +806,29 @@
 		MlmeSetTxQuality(pEntry, CurrRateIdx, DRS_TX_QUALITY_WORST_BOUND);
 		if (CurrRateIdx != DownRateIdx)
 		{
-			pEntry->CurrTxRateIndex = DownRateIdx;
-			pEntry->LastSecTxRateChangeAction = RATE_DOWN;
+			if(pAd->bPandoraBoxSmartEn) {
+						UCHAR *pTable = pEntry->pTable;
+						CHAR Rssi;
+						RTMP_RA_LEGACY_TB *pCurrTxRate;
+			
+						Rssi = RTMPAvgRssi(pAd, &pEntry->RssiSample);
+						pCurrTxRate = PTX_RA_LEGACY_ENTRY(pTable, CurrRateIdx);
+						
+						
+						if(Rssi > -60 && pCurrTxRate->CurrMCS == 3)
+						{
+							pEntry->LastSecTxRateChangeAction = RATE_NO_CHANGE;
+						}
+						else
+						{
+							pEntry->CurrTxRateIndex = DownRateIdx;
+							pEntry->LastSecTxRateChangeAction = RATE_DOWN;
+						}
+			} else {	
+						pEntry->CurrTxRateIndex = DownRateIdx;
+						pEntry->LastSecTxRateChangeAction = RATE_DOWN;
+			}
+
 		}
 	}
 	else
diff --git a/mt76x3/ap/ap_cfg.c b/mt76x3/ap/ap_cfg.c
index c0aef63..1939903 100644
--- a/mt76x3/ap/ap_cfg.c
+++ b/mt76x3/ap/ap_cfg.c
@@ -462,6 +462,12 @@ int Set_Airplay_Enable(
 	IN	PSTRING 		arg);
 #endif /* AIRPLAY_SUPPORT */
 
+#if 1
+INT Set_PandoraBoxSmartEnable(
+	IN	PRTMP_ADAPTER	pAd, 
+	IN	PSTRING 		arg);
+#endif
+
 #ifdef DYNAMIC_VGA_SUPPORT
 INT	Set_DyncVgaEnable_Proc(RTMP_ADAPTER *pAd, RTMP_STRING *arg);
 
@@ -995,10 +1001,14 @@ static struct {
 #endif /* MBSS_SUPPORT */
 #endif /* CONFIG_AP_SUPPORT */
 
-
 #ifdef AIRPLAY_SUPPORT
 		{"airplayEnable",				Set_Airplay_Enable},
 #endif /* AIRPLAY_SUPPORT*/
+		
+#if 1
+		{"PandoraBoxSmart",		Set_PandoraBoxSmartEnable},
+#endif /* AIRPLAY_SUPPORT*/
+		
 
 	{"VcoPeriod",					Set_VcoPeriod_Proc},
 #ifdef CONFIG_SNIFFER_SUPPORT
@@ -13521,6 +13531,33 @@ INT Send_QOSMAP_Configure(
 #endif /* CONFIG_HOTSPOT_R2 */
 #endif /* CONFIG_DOT11V_WNM */
 
+#if 1
+INT Set_PandoraBoxSmartEnable(
+	IN	PRTMP_ADAPTER	pAd, 
+	IN	PSTRING 		arg)
+{
+	UINT	enable = 0/*, apidx = 0*/;
+
+   enable = simple_strtol(arg, 0, 16);
+   
+   if (enable==0x1)
+   {
+   
+       pAd->bPandoraBoxSmartEn = 1;
+       DBGPRINT(RT_DEBUG_ERROR, ("###Set to PandoraBoxSmart Function Enable..\n"));
+   }
+   else /* Disable Airplay*/
+   {
+   
+      pAd->bPandoraBoxSmartEn = 0;
+      DBGPRINT(RT_DEBUG_ERROR, ("###Set to PandoraBoxSmart Function Disable...\n"));
+   }
+
+   return TRUE;
+}
+
+#endif 
+
 #ifdef AIRPLAY_SUPPORT
 /*
 ========================================================================
diff --git a/mt76x3/include/rtmp.h b/mt76x3/include/rtmp.h
index da77ddb..7a81685 100644
--- a/mt76x3/include/rtmp.h
+++ b/mt76x3/include/rtmp.h
@@ -4755,7 +4755,7 @@ MONITOR_STRUCT monitor_ctrl;
 #ifdef USB_IOT_WORKAROUND2
 	BOOLEAN  bUSBIOTReady;
 #endif
-
+    BOOLEAN 		bPandoraBoxSmartEn;
 #ifdef AIRPLAY_SUPPORT
 /*
 * Value:
diff --git a/mt76x3/common/cmm_profile.c b/mt76x3/common/cmm_profile.c
index e030c0d..b7cba42 100644
--- a/mt76x3/common/cmm_profile.c
+++ b/mt76x3/common/cmm_profile.c
@@ -2273,6 +2273,17 @@ NDIS_STATUS	RTMPSetProfileParameters(
 			retval = RT_CfgSetCountryRegion(pAd, tmpbuf, BAND_24G);
 			DBGPRINT(RT_DEBUG_TRACE, ("CountryRegion=%d\n", pAd->CommonCfg.CountryRegion));
 		}
+#if 1
+		if(RTMPGetKeyParameter("PandoraBoxSmart", tmpbuf, 10, pBuffer, TRUE))
+		{
+
+			if(simple_strtol(tmpbuf, 0, 10) != 0)  /*Enable*/
+				pAd->bPandoraBoxSmartEn = TRUE;
+			else /*Disable*/
+				pAd->bPandoraBoxSmartEn = FALSE;
+			DBGPRINT(RT_DEBUG_ERROR, ("PandoraBox SmartFunction is %s!!!\n", pAd->bPandoraBoxSmartEn?"enabled":"disabled"));
+		}
+#endif
 #ifdef P2P_CHANNEL_LIST_SEPARATE
 		/*P2PCountryRegion*/
 		if(RTMPGetKeyParameter("P2PCountryRegion", tmpbuf, 25, pBuffer, TRUE))
