--- a/mt7628/embedded/common/cmm_info.c
+++ b/mt7628/embedded/common/cmm_info.c
@@ -3214,6 +3214,8 @@ VOID RTMPIoctlGetSiteSurvey(
     INT         max_len = LINE_LEN;
 	BSS_ENTRY *pBss;
 	UINT32 TotalLen, BufLen = IW_SCAN_MAX_DATA;
+	BSS_TABLE *pScanTab;
+	pScanTab = &pAdapter->ScanTab;
 #ifdef AIRPLAY_SUPPORT
 	UCHAR TargetSsid[MAX_LEN_OF_SSID+1];
 	UCHAR TargetSsidLen = 0;
@@ -3291,6 +3293,8 @@ VOID RTMPIoctlGetSiteSurvey(
 	pAdapter->StaCfg.bSkipAutoScanConn = TRUE;
 #endif /* CONFIG_STA_SUPPORT */
 
+	BssTableSortByRssi(pScanTab,FALSE);
+
 	while ((ScanRunning(pAdapter) == TRUE) && (WaitCnt++ < 200))
 		OS_WAIT(500);
 
